File: Scala/src/main/scala/AlgorithmicHeights/HS.scala
package AlgorithmicHeights

object HS {
  import scala.annotation.tailrec

  private def convertToIntArray(line: String): Array[Int] = line.split(" ").map(_.toInt)

  private def getIndexOfParentChildrenMaximum(array: Array[Int], parentIx: Int, size: Int): Int = {
    val leftChildIx: Int = 2 * parentIx + 1
    val rightChildIx: Int = leftChildIx + 1
    val maxIndex: Int = if (leftChildIx < size && array(leftChildIx) > array(parentIx)) leftChildIx else parentIx
    if (rightChildIx < size && array(rightChildIx) > array(maxIndex)) rightChildIx else maxIndex
  }

  private def swapElems(array: Array[Int], ix: Int, jy: Int): Unit = {
    val elem: Int = array(ix)
    array(ix) = array(jy)
    array(jy) = elem
  }

  private def siftDown(array: Array[Int], parentIx: Int, n: Int): Unit = {
    @tailrec
    def loop(currentParentIx: Int): Unit = {
      val maxIx: Int = getIndexOfParentChildrenMaximum(array, currentParentIx, n)
      if (maxIx == currentParentIx) ()
      else {
        swapElems(array, maxIx, currentParentIx)
        loop(maxIx)
      }
    }

    loop(parentIx)
  }

  private def heapify(array: Array[Int], n: Int): Unit =
    (n / 2 - 1 to 0 by -1).foreach(parentIx => siftDown(array, parentIx, n))

  def heapsort(array: Array[Int], size: Int): Unit = {
    heapify(array, size)
    (size until 1 by -1).foreach{
      n =>
        swapElems(array, 0, n - 1)
        siftDown(array, 0, n - 1)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val n: Int = reader.next().toInt
    val array: Array[Int] = convertToIntArray(reader.next())
    heapify(array, n)
    println(array.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/hs.ml
let convert_to_ints (line : string) : int array =
    Array.of_list @@ List.map int_of_string Str.(line |> split (regexp " "))


let get_index_of_parent_children_maximum (arr : int array) (parent_ix : int) (size : int) : int =
    let left_child_ix = 2 * parent_ix + 1 in
    let right_child_ix = left_child_ix + 1 in
    let max_ix = if left_child_ix < size && arr.(left_child_ix) > arr.(parent_ix) then left_child_ix else parent_ix in
    if right_child_ix < size && arr.(right_child_ix) > arr.(max_ix) then right_child_ix
    else max_ix


let swap arr i j =
    let elem = arr.(i) in
    arr.(i) <- arr.(j);
    arr.(j) <- elem


let sift_down (arr : int array) (parent_ix : int) (n : int) : unit =
    let rec loop current_parent_ix =
        let max_ix = get_index_of_parent_children_maximum arr current_parent_ix n in
        if max_ix <> current_parent_ix then
            (swap arr max_ix current_parent_ix;
            loop max_ix) in
    loop parent_ix


let heapify (arr : int array) (n : int) : unit =
    for parent_ix = (n / 2 - 1) downto 0 do
        sift_down arr parent_ix n;
    done


let heapsort (arr : int array) (n : int) : unit =
    heapify arr n;
    for k = n downto 2 do
        swap arr 0 (k - 1);
        sift_down arr 0 (k - 1)
    done


let () =
    let n = read_int () in
    let arr = convert_to_ints (read_line ()) in
    heapsort arr n;
    Array.(arr |> map string_of_int |> to_list) |> String.concat " " |> print_endline
